src/native/directory_traversal.c

```diff
@@ -0,0 +1,85 @@
+/**
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+#include "crt.h"
+#include "java_class_ids.h"
+#include "retry_utils.h"
+#include <aws/common/file.h>
+#include <aws/common/string.h>
+#include <jni.h>
+
+/* on 32-bit platforms, casting pointers to longs throws a warning we don't need */
+#if UINTPTR_MAX == 0xffffffff
+#    if defined(_MSC_VER)
+#        pragma warning(push)
+#        pragma warning(disable : 4305) /* 'type cast': truncation from 'jlong' to 'jni_tls_ctx_options *' */
+#        pragma warning(disable : 4221)
+#    else
+#        pragma GCC diagnostic push
+#        pragma GCC diagnostic ignored "-Wpointer-to-int-cast"
+#        pragma GCC diagnostic ignored "-Wint-to-pointer-cast"
+#    endif
+#endif
+
+struct directory_traversal_callback_ctx {
+    JNIEnv *env;
+    jobject handler;
+};
+
+static bool s_on_directory_entry(const struct aws_directory_entry *entry, void *user_data) {
+
+    struct directory_traversal_callback_ctx *ctx = user_data;
+    JNIEnv *env = ctx->env;
+
+    jstring path = aws_jni_string_from_cursor(env, &entry->path);
+    jstring relativePath = aws_jni_string_from_cursor(env, &entry->relative_path);
+
+    jboolean callback_result = (*env)->CallBooleanMethod(
+        env,
+        ctx->handler,
+        directory_traversal_handler_properties.on_directory_entry_method_id,
+        path,
+        relativePath,
+        (jboolean)entry->file_type & AWS_FILE_TYPE_DIRECTORY,
+        (jboolean)entry->file_type & AWS_FILE_TYPE_SYM_LINK,
+        (jboolean)entry->file_type & AWS_FILE_TYPE_FILE,
+        (jlong)entry->file_size);
+
+    return (bool)callback_result;
+}
+
+JNIEXPORT void JNICALL Java_software_amazon_awssdk_crt_io_DirectoryTraversal_crtTraverse(
+    JNIEnv *env,
+    jclass jni_class,
+    jstring path,
+    jboolean recursive,
+    jobject handler) {
+    (void)jni_class;
+
+    struct aws_allocator *allocator = aws_jni_get_allocator();
+
+    const char *path_c_str = (*env)->GetStringUTFChars(env, path, NULL);
```

Comment:
we have a helper function aws_jni_new_string_from_jstring()
also, need to check result for NULL
not sure if you ned to raise a java exception before returning, or if it's already thrown by the underlying JNI function

Reply:
Beautiful, thanks!

Looked into other usages of the helper function, need to raise the Java exception before returning.